packages/@lwc/engine/src/framework/vm.ts

```diff
@@ -512,15 +511,23 @@ function recursivelyDisconnectChildren(vnodes: VNodes) {
     }
 }
 
-// This is a super optimized mechanism to remove the content of the shadowRoot
-// without having to go into snabbdom. Especially useful when the reset is a consequence
-// of an error, in which case the children VNodes might not be representing the current
-// state of the DOM
+// This is a super optimized mechanism to remove the content of the shadowRoot without having to go
+// into snabbdom. Especially useful when the reset is a consequence of an error, in which case the
+// children VNodes might not be representing the current state of the DOM
 export function resetShadowRoot(vm: VM) {
+    const { children, cmpRoot } = vm;
+
+    for (let i = 0, len = children.length; i < len; i++) {
+        const child = children[i];
+
+        if (!isNull(child) && !isUndefined(child.elm)) {
+            cmpRoot.removeChild(child.elm);
+        }
+    }
     vm.children = EmptyArray;
-    ShadowRootInnerHTMLSetter.call(vm.cmpRoot, '');
-    // disconnecting any known custom element inside the shadow of the this vm
+
     runShadowChildNodesDisconnectedCallback(vm);
+    vm.velements = EmptyArray;
```

Comment:
Is this something we were previously forgetting to do?

Reply:
I think so, resetting the shadow root should also reset all the properties that have been set during the previous rendering cycle.

Reply:
The velements are reset [before template evaluation](https://github.com/salesforce/lwc/blob/master/packages/@lwc/engine/src/framework/template.ts#L211) too. You could eliminate that.

I had a deja-vu moment because we have the exact same change in node reactions [PR](https://github.com/salesforce/lwc/pull/1798/files#diff-1643410c4aa782369b8a7912d22207aeR113) 😉